protocols/kad/src/behaviour.rs

```diff
@@ -2419,7 +2424,28 @@ where
 
     fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
         self.listen_addresses.on_swarm_event(&event);
-        self.external_addresses.on_swarm_event(&event);
+        let external_addresses_changed = self.external_addresses.on_swarm_event(&event);
+
+        if external_addresses_changed && !self.inbound_connections.is_empty() {
+            let num_connections = self.inbound_connections.len();
+
+            log::debug!(
+                "External addresses changed, re-configuring {} inbound connection{}",
+                num_connections,
+                if num_connections > 1 { "s" } else { "" }
+            );
+
+            self.queued_events
+                .extend(self.inbound_connections.iter().map(|(conn_id, peer_id)| {
+                    ToSwarm::NotifyHandler {
+                        peer_id: *peer_id,
+                        handler: NotifyHandler::One(*conn_id),
+                        event: KademliaHandlerIn::ReconfigureMode {
+                            external_addresses: self.external_addresses.iter().cloned().collect(),
+                        },
+                    }
+                }));
```

Comment:
Why do we only reconfigure the mode of inbound connections?

The remote node of an outbound connection might be aware of our external address through some other means, e.g. `libp2p-identify`. Thus, us switching to server mode, is of interest to the remote, as they can now add us to their routing table.

Reply:
How would the remote know whether we are reusing ports and are thus reachable on this observed address?

Are you arguing that the policy of not adding outbound connections to the routing table is internal to `rust-libp2p` and should not be assumed for other nodes?

Reply:
> Are you arguing that the policy of not adding outbound connections to the routing table is internal to `rust-libp2p` and should not be assumed for other nodes?

Yes, that is my understanding.